driver-core/src/main/com/mongodb/internal/operation/BulkWriteBatch.java

```diff
@@ -279,17 +277,34 @@ public FieldNameValidator getFieldNameValidator() {
 
     private BulkWriteResult getBulkWriteResult(final BsonDocument result) {
         int count = result.getNumber("n").intValue();
+        List<BulkWriteInsert> insertedItems = getInsertedItems(result);
         List<BulkWriteUpsert> upsertedItems = getUpsertedItems(result);
-        return BulkWriteResult.acknowledged(batchType, count - upsertedItems.size(), getModifiedCount(result), upsertedItems);
+        return BulkWriteResult.acknowledged(batchType, count - upsertedItems.size(), getModifiedCount(result), upsertedItems,
+                insertedItems);
     }
 
-    @SuppressWarnings("unchecked")
+    private List<BulkWriteInsert> getInsertedItems(final BsonDocument result) {
+        if (payload.getPayloadType() == SplittablePayload.Type.INSERT) {
+
+            Stream<WriteRequestWithIndex> writeRequests = payload.getWriteRequestWithIndexes().stream();
+            List<Integer> writeErrors = getWriteErrors(result).stream().map(BulkWriteError::getIndex).collect(Collectors.toList());
+            if (!writeErrors.isEmpty()) {
+                writeRequests = writeRequests.filter(wr -> !writeErrors.contains(wr.getIndex()));
+            }
+            return writeRequests
+                    .map(wr -> new BulkWriteInsert(wr.getIndex(), ((InsertRequest) wr.getWriteRequest()).getId()))
+                    .collect(Collectors.toList());
+        }
+        return Collections.emptyList();
+    }
+
+
     private List<BulkWriteUpsert> getUpsertedItems(final BsonDocument result) {
         BsonArray upsertedValue = result.getArray("upserted", new BsonArray());
-        List<BulkWriteUpsert> bulkWriteUpsertList = new ArrayList<BulkWriteUpsert>();
+        List<BulkWriteUpsert> bulkWriteUpsertList = new ArrayList<>();
         for (BsonValue upsertedItem : upsertedValue) {
             BsonDocument upsertedItemDocument = (BsonDocument) upsertedItem;
-            bulkWriteUpsertList.add(new BulkWriteUpsert(upsertedItemDocument.getNumber("index").intValue(),
+            bulkWriteUpsertList.add(new BulkWriteUpsert(indexMap.map(upsertedItemDocument.getNumber("index").intValue()),
```

Comment:
Use the indexMap to set the correct index value when creating the `BulkWriteUpsert`  - means we only create it once now.

Reply:
Got it.